src/opm/input/eclipse/Schedule/UDQ/UDQState.cpp

```diff
@@ -17,201 +17,355 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-
-#include <stdexcept>
-
 #include <opm/input/eclipse/Schedule/UDQ/UDQState.hpp>
+
 #include <opm/io/eclipse/rst/state.hpp>
 
-namespace Opm {
+#include <cstddef>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <unordered_map>
+
+#include <fmt/format.h>
 
 namespace {
 
-bool is_udq(const std::string& key) {
-    if (key.size() < 2)
-        return false;
+template <typename K, typename V>
+using UMap = std::unordered_map<K, V>;
 
-    if (key[1] != 'U')
-        return false;
+template <typename V>
+using SMap = UMap<std::string, V>;
+
+template <typename T>
+using S2Map = SMap<SMap<T>>;
+
+template <typename K, typename V>
+using SKMap = SMap<UMap<K, V>>;
 
-    return true;
+template <typename K, typename V>
+using S2KMap = S2Map<UMap<K, V>>;
+
+bool is_udq(const std::string& key)
+{
+    return (key.size() >= std::string::size_type{2})
+        && (key[1] == 'U');
 }
 
-bool has_var(const std::unordered_map<std::string, std::unordered_map<std::string, double>>& values,
-             const std::string& wgname,
-             const std::string udq_key) {
+bool has_var(const S2Map<double>& values,
+             const std::string&   wgname,
+             const std::string&   udq_key)
+{
     auto res_iter = values.find(udq_key);
-    if (res_iter == values.end())
-        return false;
+    return (res_iter != values.end())
+        && (res_iter->second.count(wgname) > 0);
+}
+
+void add_undefined_results(const Opm::UDQScalar& result,
+                           SMap<double>&         values)
+{
+    values.erase(result.wgname());
+}
+
+void add_undefined_results(const Opm::UDQScalar&       result,
```

Comment:
Should we change the function name? Does not really add anything. Maybe undefine_results would be more clear?

Reply:
> Should we change the function name? [...} Maybe `undefine_results()` would be more clear?

Good idea, I've updated the PR accordingly.